### ISSUE TITLE
Add text font/language

### ISSUE THREAD

**Issue by LittleRichard** [LittleRichard/sormtger#3]
Add a custom font/language to tesseract using the method described by:
http://michaeljaylissner.com/posts/2012/02/11/adding-new-fonts-to-tesseract-3-ocr-engine/

These files can/will need to be committed to this repo as static
* doc/pdf
* final traineddata, packaged and installed with the repo during an update.

Make 3 doc/pdf/tiffs, one for each of:
* original font
* beleren font
* beleren small caps font

Instead of using the 'training data' file from the link in this ticket, which has all kinds of non-alphanumeric characters, focus on realistic stuff for what it will be reading.  All files should include the same base cases, like:
* The alphabet upper and lower plus some punctuation
* quick-brown-fox title-cased
* pick 50+ card names and include them (title-cased)
  * the weirder, the better.

Make sure there's plenty of space in the test set document to add more text when we find cases of character groups that the OCR doesn't do well.  Since box files are a pain we should be able to:
* add the character groups to the doc/pdf/tiff
* generate a fresh box file
* cut and paste the 'new data' section of the box file to the old one
* do the box file optimization on ONLY the new stuff
* regenerate the training set

DO NOT add character spacing/padding, let them overlap and instead define character groups where 2 or more characters overlap into a single 'box'.  This will be especially helpful for the original font, where the characters connect like cursive sometimes.

**Comment by LittleRichard** [comment]
not necessary, stock english is fine

**Comment by LittleRichard** [comment]
Just kidding.

**Comment by LittleRichard** [comment]
@Panzer23 , in good news I did a very rough tesseract training on the original font and it performed only slightly worse than the stock 'english' language.  I think with some of the tricks I've learned about OCR and training along the way, I can boost it significantly by packaging the 3 fonts into a single language 'traineddata'.  Between this, camera/lighting improvements, and the database text-similarity-index, I'm optimistic that we can get the name recognition accuracy well into the upper 90s.

**Comment by LittleRichard** [comment]
@Panzer23 , last night I trained a richer set on the original font and my trained-data didn't do significantly better than the default 'english' trained-data.  It picked up some of the weirder cases, but still whiffed on a lot of them.  I've noticed that the OCR has 2 main steps; identify symbol, recognize character.  The cases that are failing are mostly NOT issues with the character recognition; it's returning None/empty-string against a well-cropped image of valid text.

We've already discussed the lighting issues, but I think the image skew is a bigger problem than I had originally thought.  It seems that even being off by a couple degrees can make it go from 'Seourgc of Heets' to empty-string.  Looks like tesseract agrees:
https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality#image-processing

That doc runs through a bunch of optimization options, which will likely be helpful, but I can't do a ton about the image rotation.  Before you go wild on the pic-tray today i figured i'd pass along that 

I'm still glad I trained the set because it's nice to have so we can test it if necessary, but i'm lukewarm again on the outlook.

**Comment by Panzer23** [comment]
@LittleRichard lukewarm on the outlook? hopefully you mean for this batch.  I know the issues surrounding the skewing, and they are all to do with the tolerances in the V0.05 version of this machine. the part that is connecting the servo to the legos is not quite perfect, and my job of editing the servo to hold the part by friction is also....less than ideal.  This means that the pictray is not perfectly lined up with the rest of the machine, most notably the camera.  in V0.1, the tray will be aligned much better with the camera/rest of the machine, and hopefully it will be less awkward.  Do not worry yourself too much at this juncture about the skewed text.  To use a joe analogy: The Newton-Raphson method for this machine has only been applied once.  It is on the docket to fix this issue.

**Comment by LittleRichard** [comment]
Noooo just lukewarm on the point of this ticket, the custom-font training I did yielding significantly better results than the stock 'english' set.  Dragging boxes and fixing H->A for a couple hours to get no result sucks, but I guess we'll call it a tie for the time you took to take all those low-light pictures haha.  Still useful to eliminate a variable, learn something, and an important step on the path, but doesn't feel great right after running the test script.

The more I think about it and look at the test crops, the more I'm convinced the skew was the source of most of the issues.  The cases where it detected text that was so wrong that it couldn't make a confident DB fuzzy-string-match were pretty infrequent, the vast majority of the failures were the result of not identifying any text at all.  Between skew fixing, more light, and some image processing cuteness, I think we have plenty of fixes for the problem.

I'm genuinely not concerned about the hardware aspect, and am definitely optimistic that it will eventually be good but not as sure if it will be excellent.  The scikit-image stuff should let us get cute with software to correct inconsistencies, so there's some flexibility there.

I'm going to close this ticket for now, I'll hold onto the trained data sets but won't commit anything to the repo around this.
